README.rst
.. image:: https://www.herokucdn.com/deploy/button.svg
   :target: https://heroku.com/deploy?template=https://github.com/adbenitez/simplebot_downloader/tree/heroku/


"Downloader" SimpleBot Plugin
=============================

.. image:: https://img.shields.io/pypi/v/simplebot_downloader.svg
   :target: https://pypi.org/project/simplebot_downloader

.. image:: https://img.shields.io/pypi/pyversions/simplebot_downloader.svg
   :target: https://pypi.org/project/simplebot_downloader

.. image:: https://pepy.tech/badge/simplebot_downloader
   :target: https://pepy.tech/project/simplebot_downloader

.. image:: https://img.shields.io/pypi/l/simplebot_downloader.svg
   :target: https://pypi.org/project/simplebot_downloader

.. image:: https://github.com/adbenitez/simplebot_downloader/actions/workflows/python-ci.yml/badge.svg
   :target: https://github.com/adbenitez/simplebot_downloader/actions/workflows/python-ci.yml

.. image:: https://img.shields.io/badge/code%20style-black-000000.svg
   :target: https://github.com/psf/black

A `SimpleBot`_ plugin that allows to download files by email/DeltaChat.

Install
-------

To install run::

  pip install simplebot-downloader

Customization
-------------

By default the maximum file size allowed is 300MB, to change it, pass the number of bytes as follows::

  simplebot -a bot@example.com db -s simplebot_downloader/max_size 10485760

By default the file will be sent in several emails split in chunks of 15728640 Bytes (15MB), to change it::

  simplebot -a bot@example.com db -s simplebot_downloader/part_size 10240

By default there is a delay of 60 seconds between each sent chunk, to change it::

  simplebot -a bot@example.com db -s simplebot_downloader/delay 120

By default the bot is in "filter" mode, this mean to queue downloads you just send the file link directly, to download with a command instead::

  simplebot -a bot@example.com db -s simplebot_downloader/mode command


.. _SimpleBot: https://github.com/simplebot-org/simplebot


CHANGELOG.rst
Changelog
*********

`Unreleased`_
-------------

- changed default `max_size` to 300MB.
- set part number as display name of the message to preview before downloading

1.0.0
-----

- initial release


.. _Unreleased: https://github.com/adbenitez/simplebot_downloader/compare/v1.0.0...HEAD


tests/test_plugin.py
class TestPlugin:
    def test_filter(self, mocker) -> None:
        mocker.get_one_reply("https://delta.chat")

        assert not mocker.get_replies("message without link")


heroku.sh
#!/bin/bash

# configure the bot
python3 -m simplebot init "$ADDR" "$PASSWORD"
python3 -m simplebot -a "$ADDR" set_name "File Downloader"

# add the youtube plugin
python3 -m pip install youtube-dl # required by youtube.py plugin
python3 -c "import requests; r=requests.get('https://github.com/adbenitez/simplebot-scripts/raw/master/scripts/youtube.py'); open('youtube.py', 'wb').write(r.content)"
python3 -m simplebot -a "$ADDR" plugin --add ./youtube.py

# add the encryption_error plugin to leverage key changes
python3 -c "import requests; r=requests.get('https://github.com/adbenitez/simplebot-scripts/raw/master/scripts/encryption_error.py'); open('encryption_error.py', 'wb').write(r.content)"
python3 -m simplebot -a "$ADDR" plugin --add ./encryption_error.py

# add admin plugin
if [ -n "$ADMIN" ]; then
    python3 -m pip install psutil # required by admin.py plugin
    python3 -c "import requests; r=requests.get('https://github.com/adbenitez/simplebot-scripts/raw/master/scripts/admin.py'); open('admin.py', 'wb').write(r.content)"
    python3 -m simplebot -a "$ADDR" plugin --add ./admin.py
    python3 -m simplebot -a "$ADDR" admin --add "$ADMIN"
fi

# start the bot
python3 -m simplebot -a "$ADDR" serve


simplebot_downloader/util.py
"""Utilities."""

import functools
import mimetypes
import os
import re
from tempfile import TemporaryDirectory
from typing import Callable, Generator

import multivolumefile
import py7zr
import requests
from simplebot.bot import DeltaBot

session = requests.Session()
session.headers.update(
    {
        "user-agent": "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:60.0) Gecko/20100101 Firefox/60.0"
    }
)
session.request = functools.partial(session.request, timeout=15)  # noqa


class FileTooBig(ValueError):
    """File is too big."""


def get_setting(bot: DeltaBot, key: str, value=None) -> str:
    """Get setting value, if value is given and the setting is not set, the setting will be set to the given value."""
    scope = __name__.split(".", maxsplit=1)[0]
    val = bot.get(key, scope=scope)
    if val is None and value is not None:
        bot.set(key, value, scope=scope)
        val = value
    return val


def download_file(url: str, folder: str, max_size: int) -> str:
    """Download URL and save the file in the give folder.

    If the file is bigger than max_size a FileTooBig exception is raised.
    Returns the path of the downloaded file.
    """
    if not url.startswith("http"):
        url = "http://" + url
    with session.get(url, stream=True) as resp:
        resp.raise_for_status()
        filepath = os.path.join(folder, get_filename(resp)[-20:].lstrip("."))
        with open(filepath, "wb") as file:
            size = 0
            for chunk in resp.iter_content(chunk_size=1024 * 500):
                size += len(chunk)
                if size > max_size:
                    raise FileTooBig(
                        f"Only files smaller than {sizeof_fmt(max_size)} are allowed"
                    )
                file.write(chunk)

    return filepath


def split_download(
    url: str, part_size: int, max_size: int, downloader: Callable = download_file
) -> Generator:
    with TemporaryDirectory() as tempdir:
        path = downloader(url, tempdir, max_size)
        if os.stat(path).st_size > part_size:
            with multivolumefile.open(path + ".7z", "wb", volume=part_size) as vol:
                with py7zr.SevenZipFile(
                    vol, "w", filters=[{"id": py7zr.FILTER_COPY}]
                ) as archive:
                    archive.write(path, os.path.basename(path))

            os.remove(path)
            parts = sorted(os.listdir(tempdir))
            parts_count = len(parts)
            for num, filename in enumerate(parts, 1):
                path = os.path.join(tempdir, filename)
                yield path, num, parts_count
                os.remove(path)
        else:
            yield path, 1, 1


def get_filename(resp: requests.Response) -> str:
    disp = resp.headers.get("content-disposition")
    if disp is not None and re.findall("filename=(.+)", disp):
        fname = re.findall("filename=(.+)", disp)[0].strip('"')
    else:
        fname = resp.url.split("/")[-1].split("?")[0].split("#")[0]

    if "." in fname:
        return fname

    ctype = resp.headers.get("content-type", "").split(";")[0].strip().lower()
    if ctype == "text/plain":
        ext = ".txt"
    elif ctype == "image/jpeg":
        ext = ".jpg"
    else:
        ext = mimetypes.guess_extension(ctype) or ""
    return (fname or "file") + ext


def sizeof_fmt(num: float) -> str:
    """Format size in human redable form."""
    suffix = "B"
    for unit in ["", "Ki", "Mi", "Gi", "Ti", "Pi", "Ei", "Zi"]:
        if abs(num) < 1024.0:
            return "%3.1f%s%s" % (num, unit, suffix)  # noqa
        num /= 1024.0
    return "%.1f%s%s" % (num, "Yi", suffix)  # noqa


simplebot_downloader/__init__.py
"""Hooks and filters."""

import time
from threading import Thread
from typing import Callable, Dict, Generator

import simplebot
from deltachat import Message
from simplebot.bot import DeltaBot, Replies

from .util import FileTooBig, download_file, get_setting, split_download

DEF_MAX_SIZE = str(1024**2 * 300)
DEF_PART_SIZE = str(1024**2 * 15)
DEF_DELAY = "60"
MAX_QUEUE_SIZE = 50
downloads: Dict[str, Generator] = {}


@simplebot.hookimpl
def deltabot_init(bot: DeltaBot) -> None:
    get_setting(bot, "max_size", DEF_MAX_SIZE)
    get_setting(bot, "part_size", DEF_PART_SIZE)
    get_setting(bot, "delay", DEF_DELAY)
    mode = get_setting(bot, "mode", "filter")

    if mode == "filter":
        bot.filters.register(download_filter)
    else:
        bot.commands.register(download_cmd, name="/download")


@simplebot.hookimpl
def deltabot_start(bot: DeltaBot) -> None:
    Thread(target=_send_files, args=(bot,)).start()


def download_filter(bot: DeltaBot, message: Message, replies: Replies) -> None:
    """Send me in private any direct download link and I will send you the file.

    Example:
    https://example.com/path/to/file.zip
    """
    if message.chat.is_multiuser() or not message.text.startswith("http"):
        return
    queue_download(message.text, bot, message, replies)


def download_cmd(
    bot: DeltaBot, payload: str, message: Message, replies: Replies
) -> None:
    """Download the given file link.

    Example:
    /download https://example.com/path/to/file.zip
    """
    queue_download(payload, bot, message, replies)


def queue_download(
    url: str,
    bot: DeltaBot,
    message: Message,
    replies: Replies,
    downloader: Callable = download_file,
) -> None:
    addr = message.get_sender_contact().addr
    if addr in downloads:
        replies.add(text="❌ You already have a download in queue", quote=message)
    elif len(downloads) >= MAX_QUEUE_SIZE:
        replies.add(
            text="❌ I'm too busy with too many downloads, try again later",
            quote=message,
        )
    else:
        replies.add(text="✔️ Request added to queue", quote=message)
        part_size = int(get_setting(bot, "part_size"))
        max_size = int(get_setting(bot, "max_size"))
        downloads[addr] = split_download(url, part_size, max_size, downloader)


def _send_files(bot: DeltaBot) -> None:
    replies = Replies(bot, bot.logger)
    while True:
        items = list(downloads.items())
        bot.logger.debug("Processing downloads queue (%s)", len(items))
        start = time.time()
        for addr, parts in items:
            chat = bot.get_chat(addr)
            try:
                path, num, parts_count = next(parts)
                text = f"Part {num}/{parts_count}"
                replies.add(text=text, sender=text, filename=path, chat=chat)
                replies.send_reply_messages()
                if num == parts_count:
                    next(parts, None)  # close context
                    downloads.pop(addr, None)
            except FileTooBig as ex:
                downloads.pop(addr, None)
                replies.add(text=f"❌ {ex}", chat=chat)
                replies.send_reply_messages()
            except (StopIteration, Exception) as ex:
                bot.logger.exception(ex)
                downloads.pop(addr, None)
                replies.add(
                    text="❌ Failed to download file, is the link correct?", chat=chat
                )
                replies.send_reply_messages()
        delay = int(get_setting(bot, "delay")) - time.time() + start
        if delay > 0:
            time.sleep(delay)
